test/e2e/network/dual_stack.go

```diff
@@ -0,0 +1,308 @@
+				},
+				Template: v1.PodTemplateSpec{
+					ObjectMeta: metav1.ObjectMeta{
+						Labels: map[string]string{"test": "dualstack-test-client"},
+					},
+					Spec: v1.PodSpec{
+						Containers: []v1.Container{
+							{
+								Name:    "dualstack-test-client",
+								Image:   networkUtilImage,
+								Command: []string{"/bin/sh", "-c", "--"},
+								Args:    []string{"while true; do sleep 30; done;"},
+							},
+						},
+						Affinity: &v1.Affinity{
+							PodAntiAffinity: &v1.PodAntiAffinity{
+								RequiredDuringSchedulingIgnoredDuringExecution: []v1.PodAffinityTerm{
+									{
+										LabelSelector: &metav1.LabelSelector{
+											MatchExpressions: []metav1.LabelSelectorRequirement{
+												{
+													Key:      "test",
+													Operator: metav1.LabelSelectorOpIn,
+													Values:   []string{"dualstack-test-client"},
+												},
+											},
+										},
+										TopologyKey: "kubernetes.io/hostname",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		}
+
+		serverDeployment, err := cs.AppsV1().Deployments(f.Namespace.Name).Create(serverDeploymentSpec)
+		framework.ExpectNoError(err)
+
+		clientDeployment, err := cs.AppsV1().Deployments(f.Namespace.Name).Create(clientDeploymentSpec)
+		framework.ExpectNoError(err)
+
+		err = e2edeploy.WaitForDeploymentComplete(cs, serverDeployment)
+		framework.ExpectNoError(err)
+		err = e2edeploy.WaitForDeploymentComplete(cs, clientDeployment)
+		framework.ExpectNoError(err)
+
+		serverPods, err := e2edeploy.GetPodsForDeployment(cs, serverDeployment)
+		framework.ExpectNoError(err)
+
+		clientPods, err := e2edeploy.GetPodsForDeployment(cs, clientDeployment)
+		framework.ExpectNoError(err)
+
+		assertNetworkConnectivity(f, *serverPods, *clientPods)
+	})
+})
+
+var (
+	duration     = "10s"
+	pollInterval = "1s"
+	timeout      = 10
+)
+
+func assertNetworkConnectivity(f *framework.Framework, serverPods v1.PodList, clientPods v1.PodList) {
+	// curl from each client pod to all server pods to assert connectivity
+	var serverIPs []string
+	for _, pod := range serverPods.Items {
+		if pod.Status.PodIPs == nil || len(pod.Status.PodIPs) != 2 || !isIPv6(pod.Status.PodIPs[1].IP) {
+			framework.Failf("PodIPs list not expected value, got %v", pod.Status.PodIPs)
+		}
+		serverIPs = append(serverIPs, pod.Status.PodIPs[0].IP, pod.Status.PodIPs[1].IP)
+	}
+
+	for _, clientPod := range clientPods.Items {
+		for _, ip := range serverIPs {
+			gomega.Consistently(func() error {
+				ginkgo.By(fmt.Sprintf("checking connectivity from pod %s to server ip %s", clientPod.Name, ip))
+				cmd := checkNetworkConnectivity(ip)
+				_, _, err := f.ExecCommandInContainerWithFullOutput(clientPod.Name, "dualstack-test-client", cmd...)
+				return err
+			}, duration, pollInterval).ShouldNot(gomega.HaveOccurred())
+		}
+	}
+}
+
+func checkNetworkConnectivity(ip string) []string {
+	curl := fmt.Sprintf("curl -g --connect-timeout %v http://[%s]", 10, ip)
```

Comment:
I suggest using https://golang.org/pkg/net/#JoinHostPort so it will work for ipv4 and ipv6

`curl := fmt.Sprintf("curl -g --connect-timeout %v http://%s", 10, net.JoinHostPort(ip, 80))

Reply:
@aojea The curl command I've above works fine for ipv4 and ipv6. But I think what you suggested looks clean. I'll make the change. Thanks for the suggestion.